hibernate-core/src/main/java/org/hibernate/dialect/function/OrdinalFunction.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.dialect.function;
+
+import java.util.List;
+
+import org.hibernate.metamodel.mapping.JdbcMapping;
+import org.hibernate.query.ReturnableType;
+import org.hibernate.query.sqm.function.AbstractSqmSelfRenderingFunctionDescriptor;
+import org.hibernate.query.sqm.produce.function.StandardArgumentsValidators;
+import org.hibernate.query.sqm.produce.function.StandardFunctionReturnTypeResolvers;
+import org.hibernate.sql.ast.SqlAstTranslator;
+import org.hibernate.sql.ast.spi.SqlAppender;
+import org.hibernate.sql.ast.tree.SqlAstNode;
+import org.hibernate.sql.ast.tree.expression.Expression;
+import org.hibernate.type.StandardBasicTypes;
+import org.hibernate.type.descriptor.java.EnumJavaType;
+import org.hibernate.type.descriptor.jdbc.JdbcType;
+import org.hibernate.type.spi.TypeConfiguration;
+
+import static java.lang.String.format;
+
+
+/**
+ * The HQL {@code ordinal()} function returns the ordinal value of an enum
+ * <p>
+ * For enum fields mapped as ORDINAL it's a synonym for {@code cast(x as Integer)}. Same as {@link CastStrEmulation} but for Integer.
+ * For enum fields mapped as STRING or ENUM it's a case statement that returns the ordinal value.
+ *
+ * @author Luca Molteni
+ */
+public class OrdinalFunction
+		extends AbstractSqmSelfRenderingFunctionDescriptor {
+
+	public OrdinalFunction(TypeConfiguration typeConfiguration) {
+		super(
+				"ordinal",
+				StandardArgumentsValidators.exactly( 1 ),
```

Comment:
It would be nice if you could extend `FunctionParameterType` by a new constant `ENUM` and `org.hibernate.query.sqm.produce.function.ArgumentTypesValidator#isCompatible` like I have done in the past for e.g. `IMPLICIT_JSON`. You can then validate that the passed argument really is an enum.

Reply:
I pushed something that should resemble what you've done with `IMPLICIT_JSON`

The only thing I don't like that is there are now two `ENUM` types, one that is actually `ENUM` and one that is implicit enum